### ISSUE TITLE
Property in subset of branches of a union type should be accessible for checking against undefined and result in narrowing type of object

### ISSUE THREAD

**Issue by callionica** [microsoft/TypeScript#49933]
## Update
I opened this suggestion before learning that `in` does the narrowing that I was looking for (thanks to fatcerberus).

Maybe this suggestion should become one for an updated error message or a quick fix that converts `if (u.p !== undefined)` to `if ("p" in u)` when `u` is a union and `p` is a property only available in a subset of union members.

-----

# Suggestion

## 🔍 Search Terms

undefined, subset property, union, narrowing, type predicate, undefined test

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [X] This wouldn't change the runtime behavior of existing JavaScript code
* [X] This could be implemented without emitting different JS based on the types of the expressions
* [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [X] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

TLDR: Type predicates are too much ceremony when a simple check against `undefined` will do

Details: 
Currently, when you have a union of types such as `({ a: number } | { b: string })`, TS prevents you from writing this natural code:

```
if (x.a !== undefined) { // <-- Compiler error here...
    x; // <-- ... but it should work so that compiler can see x as { a: number } here
}
```

It's annoying when you have just told the compiler that a property might exist, that you can't then test whether the property does exist.
It's also annoying that when you have told the compiler that a property only exists in one branch of a union, that testing the property doesn't let the compiler figure out that the object is in that branch of the union.

TS provides workarounds, but none of them are great.

You can provide a type assertion to get access to the member, but that doesn't allow the compiler to narrow the type of `x`.

Alternatively, you can use a type predicate to distinguish between parts of a union, but that means defining a separate function which interrupts your flow and still requires you to use unintuitive code.

Here's the type predicate example from the docs:

```
function isFish(pet: Fish | Bird): pet is Fish {
  return (pet as Fish).swim !== undefined;
}
```

You want to find out if something is a fish, so you, counterintuitively, have to declare that it _is_ a fish before you can find out whether it is a fish.

Unfortunately type predicates are not very reusable. For example, this predicate does not tell us that `pet` is a fish in all cases. Instead, it allows us to distinguish between a fish and a bird. If we need to detect a fish when we might have a fish, a bird, and a dog, we need to create a new function or adjust the type of the existing one.

It would be a big usability win not to have to define type predicates for situations like this.

My suggestion is that code using the following pattern should compile without error and narrow the object:

0. In a conditional
1. Operation is a property access on a union type
2. Property name exists in at least one branch of the union
3. Property value is tested equal/not-equal to `undefined`

Note that the optional chain operator also fulfills these conditions, it's not just expressions of the form `if (a.b === undefined)`.

This change would neither break existing code (such code currently does not compile) nor would it fail to detect probable errors.

There is an error that would no longer be detected: If the user intended to enter the name of a property available in all branches of the union, but typoed and instead provided the name of a property only valid in a subset of branches, they would previously get a compile error, but with the change, the property access and test would now compile. This is not one to be concerned about.

## 📃 Motivating Example

ANNOUNCEMENT:
Type predicates are needed in fewer cases, and more of your code will just work like you expect.

Checking a property against `undefined` is now allowed for union types even if the property is not present in all members of the union and the compiler will use that check to narrow the type of the object down to the appropriate member(s) of the union.

The following code will now compile successfully exactly as written:

```
type Fish = { swimming: boolean; };
type Cat = { purring: boolean; }

declare const pet: Fish | Cat;

if (pet.swimming !== undefined) {
    pet.swimming = true; // It's a fish
} else {
    pet.purring = true; // It's a cat
}
```

Previously, you would have had to introduce a type predicate to decide whether `pet` was a cat or a fish. Now you can just write code that checks the distinguishing property directly and TS will do the work of determining what types are appropriate.

## 💻 Use Cases

It's really quite common to want to distinguish between members of a union type (you're defining an overloaded function or consuming someone's API for example) and many times you do it by checking for the presence of a property that is only available in one of (or a subset of) the union's members. Currently, the obvious code to detect property presence fails the type check and TS produces an error, so you have to go through the faff of creating a type predicate that is only useful in this one situation. This change would fix that.

This current behavior is probably my least favorite thing about TypeScript.

**Comment by MartinJohns** [comment]
Duplicate of #42775.

Your suggestion would be very unsound. TypeScript types are not sealed. it's possible there are additional properties not specified in the type definition.

**Comment by callionica** [comment]
Is it a duplicate? I'm not proposing that TS treat properties as optional. This code would continue to fail for example:

```
type Fish = { swimming: boolean; };
type Cat = { purring: boolean; }

declare const pet: Fish | Cat;

pet.swimming = true; // This produces an error even if suggestion is implemented
```

Your point about soundness also applies to checking properties against `undefined` when implementing a type predicate. Do you recommend that users not check properties against `undefined` in type predicates? I'm probably missing your point. I hope you have the time to clarify.

**Comment by MartinJohns** [comment]
> I'm not proposing that TS treat properties as optional.

You're suggesting that the type you access has these "fictional" properties, at least on read access.

> Your point about soundness also applies to checking properties against `undefined` when implementing a type predicate.

In type predicates the user is responsible that the verification is correct. This is an intentional decision, your suggestion would be a global pitfall

> Do you recommend that users not check properties against `undefined` in type predicates?

And depending on the code and the data I would recommend against simply checking for `undefined` and verify your data actually matches your type, as much as possible.

> I am suggesting that names of properties that exist on a subset of a union's members are accessible only in situations where they are being compared equal or not equal to undefined.

So you're suggesting the type you're accessing a member on is changing depending on the used operator and the right (or left) side expression.

---

I'm not a member of the TypeScript team, of course. But I absolutely can't see this suggestion being a good fit for the language. Anyway, I said by bit, I'm out.

**Comment by callionica** [comment]
I appreciate the discussion, but I want to be clear what I'm proposing. I'm not suggesting that there should be read access to fictional properties. My suggestion is not to allow read access at all: it is to allow equal/not equal comparison to undefined only. (The result of that comparison then allows TS to narrow the type). The relevant properties of course aren't fictional, they are properties that are present on the members of the union.

> So you're suggesting the type you're accessing a member on is changing depending on the used operator and the right (or left) side expression.

The type on which the member is accessed is the union. Yes, the context in which the member is accessed would determine what happens in the compiler.

**Comment by fatcerberus** [comment]
You seem to be proposing this:

```ts
type Fish = { swimming: boolean; };
type Cat = { purring: boolean; }

const animal = { purring: true, swimming: "mercat" };
const pet: Fish | Cat = animal;
if (pet.swimming !== undefined) {
    pet;  // narrowed to Fish, but it's actually a Cat!
}
```

That's the unsoundness @MartinJohns is referring to.  It's worth noting that `in` checks have this problem too (and you can already check for `"swimming" in pet` to do the desired narrowing), but I wouldn't want to purposely introduce another manifestation of this same footgun.

**Comment by callionica** [comment]
Yes, that's what I'm proposing. I understand that TS allows assignment with excess properties. I'm also aware that TS allows type assertions and that objects often come from untyped JS code as `any`. There are numerous ways in which the runtime object and the types may not match.

But let's not lose sight of the fact that whereas users make mistakes, they also know more about their code than the compiler. The user wrote `if (pet.swimming !== undefined)` and expected something to happen. What do you think they meant? Non-goal 3 applies.

**Comment by fatcerberus** [comment]
> The user wrote `if (pet.swimming !== undefined)` and expected something to happen. What do you think they meant?

They clearly meant to ask "Is this a fish?"  If the type system answers "yes" (by narrowing the type to `Fish`) then it's potentially lying to them, which weakens the type system and makes it easier to easier to fall into a soundness trap without providing any additional benefits as a tradeoff.  Obviously perfect soundness isn't a design goal, but since you can already use an `in` check to achieve exactly this, I don't see how giving the user a second way to do the same unsound inference is an improvement.

**Comment by fatcerberus** [comment]
> whereas users make mistakes, they also know more about their code than the compiler

I don't mean this to be patronizing, so don't take it that way - but this is a slippery slope.  Taken to its logical conclusion, we could easily turn this into an argument to just ignore the type system entirely and write code in plain JavaScript. 😄 (Of course nobody writing in TS would actually suggest this... hopefully 😅)

Even putting soundness concerns aside, the compiler's job is to warn the user about potential mistakes, and IMO this *is* a potential mistake - the code doing the `pet.swimming` check could easily have ended up getting a `Cat` with extraneous properties without the developer who wrote the check knowing that was a possibility.  Sure, nobody's passing your function mercats *now*... but another developer may do so later under the rationale that, TS allows objects with extra properties, so any code that accepts objects of any type should be prepared to handle that properly.

**Comment by callionica** [comment]
No worries. I appreciate the discussion and agree about slippery slopes - I'm all for using the type system. The thing is that users have to be able to distinguish between types sometimes. Do you have a sound way of doing that?

I actually missed that `in` does the narrowing I'm after, so thanks much for pointing that out. I'll use that instead.

**Comment by jcalz** [comment]
The *sound* way of doing this is to change the types to
```ts
type Fish = { swimming: boolean; purring?: never };
type Cat = { purring: boolean; swimming?: never };
```
which tells the compiler that a `Fish` does not have a defined `purring` property and that ` Cat` does not have a defined `swimming` property.  After that, your original check will work:

```ts
declare const pet: Fish | Cat;
if (pet.swimming !== undefined) {
    pet.swimming = true; // It's a fish
} else {
    pet.purring = true; // It's a cat
}
```

---

If you keep it the way it is now and use [`in` operator narrowing](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-in-operator-narrowing), then the error goes away, but it has the same pitfalls as your proposal.  

It's a bit frustrating for me to have to explain both
* You can't use `obj[key] !== undefined` to narrow `obj` because that would be unsound.  Object types are not sealed or exact; what if you get some pathologically weird object from outer space that breaks your assumptions?  The world would descend into chaos!  Soundness forever!!! 
* Of course you can use `key in obj` to narrow `obj`, because it's very practical to do so.  Nobody would ever write such a check if it were remotely likely that it would fail to discriminate among union members.  Who cares if it's not sound? We've got code to write, and these seat belts are cramping our collective style!  Productivity above all!!!

without coming off as some combination of hypocritical, incompetent, and esoteric.  It would be so nice to pick one or the other philosophy here so that the explanations would at least be consistent with each other (a.k.a second-order soundness).

**Comment by fatcerberus** [comment]
> It would be so nice to pick one or the other philosophy here so that the explanations would at least be consistent with each other (a.k.a second-order soundness).

Fair point, though at least for me, the fact that `obj[key] !== undefined` doesn't work arises as a natural consequence of `obj[key]` not working *in general* because the type doesn't support the key in question.  So there's your second-order soundness.  It'd be a bit of a weird special case IMO to say you can access any arbitrary property *only for the purpose of comparing its value against `undefined`*, particularly when `key in obj` already 1) approximates that check well enough in most cases and 2) communicates intent better.

Basically, w.r.t. the aforementioned inconsistency, allowing this is just moving the goalposts: "Why can I do `obj[key] !== undefined` but not, e.g. `obj[key] === "some sentinel value"` even though both checks are potentially unsound?"

**Comment by fatcerberus** [comment]
I kind of wish we would get [Exact Types](https://github.com/microsoft/TypeScript/issues/12936) already because it does seem like lots of people *want* to write code with the expectation of no extraneous properties (and who can blame them?) and not having a way to tell the compiler to enforce that invariant is frustrating.

**Comment by callionica** [comment]
Does the `never` code work? I would have thought that after TS has seen `pet.swimming !== undefined`, it would either know that `pet.swimming` is not `undefined`, so it's something, so it's not `never`, so it's a `Fish`. Or it would know that `pet.swimming` is `undefined` and since `undefined` is not `never`, it would conclude that it's a `Fish`.

**Comment by RyanCavanaugh** [comment]
> without coming off as some combination of hypocritical, incompetent, and esoteric. 

The reason that `in` gets special treatment is that if you're writing
```ts
if (x.length) {
```
it's really not clear if you're testing `x` to *have* a length property, or if you're testing `x` to have a nonzero length property. The corresponding construction `"length" in x` has no such ambiguity of intent - it's entirely clear that what's about to follow is going to be a type assertion based on that result.

**Comment by fatcerberus** [comment]
@RyanCavanaugh The proposal here was to special-case `x.foo !== undefined` (and specifically *only* comparisons to `undefined`) as a type guard, not the truthiness-check form.

**Comment by callionica** [comment]
I'll close this as won't fix, so y'all don't spend too much time on it. Treat it as a data point. I'm satisfied with the alternative `in` test for my purposes. There is an opportunity for a better error message or quick fix here, but you've probably got higher pri stuff to deal with.
